cou/apps/base.py

```diff
@@ -909,29 +909,34 @@ def _check_application_target(self, target: OpenStackRelease) -> None:
             )
 
     def _check_mismatched_versions(self) -> None:
-        """Check that there are no mismatched versions on app units.
+        """Check that there are no unexpected mismatched versions on app units.
 
-        :raises MismatchedOpenStackVersions: When the units of the app are running
+        For cases where mismatched versions may be expected
+        (eg. nova-compute which is upgraded unit-by-unit),
+        then it will not check.
+
+        :raises MismatchedWorkloadVersions: When the units of the app are running
                                              different OpenStack versions.
         """
-        # NOTE (gabrielcocenza) nova-compute is upgraded using paused-single-unit,
-        # so it's possible to have mismatched version in applications units that are
-        # nova-compute or colocated with it.
+        # nova-compute is upgraded one unit at a time,
+        # so it's possible to have mismatched version in
+        # units of applications that are nova-compute or colocated with it.
         if any(
             "nova-compute" in app_charm
             for machine in self.machines.values()
             for app_charm in machine.apps_charms
         ):
             return
 
-        o7k_versions = self.o7k_release_units
-        if len(o7k_versions.keys()) > 1:
-            mismatched_repr = [
-                f"'{openstack_release.codename}': {units}"
-                for openstack_release, units in o7k_versions.items()
-            ]
-
-            raise MismatchedOpenStackVersions(
-                f"Units of application {self.name} are running mismatched OpenStack versions: "
-                f"{', '.join(mismatched_repr)}. This is not currently handled."
+        if len({unit.workload_version for unit in self.units.values()}) > 1:
+            formatted_results = "\n".join(
+                f"  {unit.name}: {unit.workload_version} ({self.get_latest_o7k_version(unit)})"
+                for unit in sorted(self.units.values(), key=lambda unit: unit.name)
+            )
+            raise MismatchedWorkloadVersions(
```

Comment:
You can't look at the version itself because it will break as it happens in designate-bind. The logic is right, you just need to refactor the message to show the workload version

Reply:
I'm confused - according to the conversation on https://github.com/canonical/charmed-openstack-upgrader/issues/487 : 
I asked:

> Are there any cases where workload versions may be mismatched, but the relevant openstack release is still the same (eg. if the versions were 21.0.1 and 21.0.2, only differing by a patch release)? Is it ok if workload versions are mismatched at all?

And both you and @jneo8 said that this case should not happen:

> This shouldn't happen and can be danger. @jneo8 

> This is not suppose to happen. @gabrielcocenza

Reply:
I realized that this is not the same problem I thought it should be.

Should we check if the workload version is in the version range as defined by `openstack_lookup.csv` for a given o7k release?

E.g if COU detect the app is in yoga, and the version range is [A, A+10), the we only need to make sure that all workload version of the units in that app is in this range

Reply:
> I'm confused - according to the conversation

Sorry I've interpreted wrong the question :sweat_smile: 

I thought you were asking if for example 10.0.2 would be considered different from 10.0.3 when both are considered Ussuri.

They can have different versions in the same OpenStack release before upgrading. Users can potentially ssh to a unit so troubleshoot and upgrade the package and not do the same for the other units.

Reply:
Yep ok that makes sense. So we should go back to only verifying the openstack release matches. And the work here can be simply improving the error message.

Reply:
Updated :)